softlayer/resource_softlayer_virtual_guest.go

```diff
@@ -977,61 +955,61 @@ func WaitForIPAvailable(d *schema.ResourceData, meta interface{}, private bool)
-func WaitForSecondaryIPAvailable(d *schema.ResourceData, meta interface{}) (interface{}, error) {
-	log.Printf("Waiting for server (%s) to get secondary IPs", d.Id())
-
-	stateConf := &resource.StateChangeConf{
-		Pending: []string{"unavailable"},
-		Target:  []string{"available"},
-		Refresh: func() (interface{}, string, error) {
-			fmt.Println("Refreshing secondary IPs state...")
-			secondarySubnetResult, err := services.GetAccountService(meta.(ProviderConfig).SoftLayerSession()).
-				Mask("ipAddresses[id,ipAddress]").
-				Filter(filter.Build(filter.Path("publicSubnets.endPointIpAddress.virtualGuest.id").Eq(d.Id()))).
-				GetPublicSubnets()
-			if err != nil {
-				return nil, "", err
-			}
-			if len(secondarySubnetResult) == 0 {
-				return secondarySubnetResult, "unavailable", nil
-			}
-			return secondarySubnetResult, "available", nil
-		},
-		Timeout:    45 * time.Minute,
-		Delay:      10 * time.Second,
-		MinTimeout: 10 * time.Second,
-	}
-	return stateConf.WaitForState()
-}
-
-// WaitForNoActiveTransactions Wait for no active transactions
-func WaitForNoActiveTransactions(d *schema.ResourceData, meta interface{}) (interface{}, error) {
-	log.Printf("Waiting for server (%s) to have zero active transactions", d.Id())
+// WaitForVirtualGuestAvailable Waits for virtual guest creation
+func WaitForVirtualGuestAvailable(d *schema.ResourceData, meta interface{}) (interface{}, error) {
+	log.Printf("Waiting for server (%s) to be available.", d.Id())
 	id, err := strconv.Atoi(d.Id())
 	if err != nil {
 		return nil, fmt.Errorf("The instance ID %s must be numeric", d.Id())
 	}
 
+	ipAddress := "PrimaryIpAddress"
+	if d.Get("private_network_only").(bool) {
+		ipAddress = "PrimaryBackendIpAddress"
+	}
+
 	stateConf := &resource.StateChangeConf{
-		Pending: []string{"retry", "active"},
-		Target:  []string{"idle"},
+		Pending: []string{"retry", "provisioning"},
+		Target:  []string{"available"},
 		Refresh: func() (interface{}, string, error) {
+			// Check active transactions
 			service := services.GetVirtualGuestService(meta.(ProviderConfig).SoftLayerSession())
-			transactions, err := service.Id(id).GetActiveTransactions()
+			result, err := service.Id(id).Mask("activeTransaction").GetObject()
 			if err != nil {
 				if apiErr, ok := err.(sl.Error); ok && apiErr.StatusCode == 404 {
-					return nil, "", fmt.Errorf("Couldn't get active transactions: %s", err)
+					return nil, "", fmt.Errorf("Error retrieving virtual guest: %s", err)
 				}
-
 				return false, "retry", nil
 			}
 
-			if len(transactions) == 0 {
-				return transactions, "idle", nil
+			// Check active transactions.
+			log.Println("Checking active transactions.")
+			if result.ActiveTransaction != nil {
+				return result, "provisioning", nil
 			}
 
-			return transactions, "active", nil
+			// Check Primary IP address availability.
+			log.Println("Checking Primary IP address.")
+			if sl.Grab(result, ipAddress) == "" {
```

Comment:
sl.Get will perform better as it uses less reflection. Also, ipAddress does not happen to be a nested path.